library/init/meta/environment.lean

```diff
@@ -68,7 +68,7 @@ meta constant trust_lvl       : environment → nat
 /-- Add a new declaration to the environment -/
 meta constant add             : environment → declaration → exceptional environment
 /-- make declaration `n` protected -/
-meta constant mk_protected    : environment → name → exceptional environment
+meta constant add_protected   : environment → declaration → exceptional environment
```

Comment:
```suggestion
meta constant add_protected   : environment → name → environment
```
I think you should revert this back to `add_protected : environment → name → environment` just like it is in Lean 4.

BTW, I don't think this one can fail either.  Both functions just modify the `m_protected` set without ever checking that there exist declarations with this name.  You can probably do cool things with this feature, like making declarations protected before they are declared. :smile:

Reply:
That sounds like a funny application but is there any real benefit to splitting the two? One downside I can see is carelessly protecting the wrong declaration or even protecting arbitrary definitions. I feel like this way, it's easier to use well.

Reply:
Yes, the application was a joke.  The benefit is to have the same API as the C++ code (and Lean 4).  I'm very much against designing a separate restricted special-purpose API that is also different from the C++ code.  If it is easy possible (and it is in this case), then we should always expose the C++ API as literally as we can.  We can easily build nicer abstractions on top of it in Lean.

If you want to make it easier to use, please add separate `add_protected_decl : environment → declaration → exceptional environment` and `add_protected_decl : declaration → tactic unit` functions.

Reply:
I think the case here for having a separate API in Lean and in C++ is similar to the case for having private functions in some modules. They don't provide a good way to use the module and there's no good use cases that need them.

Reply:
(here, the private stuff is the C++ code that does not get exported)

Reply:
There are advantages to having a separate `add_protected` function that just takes a name.  Namely, it composes well with other functions that create declarations.  For example, you can mark declarations created with `mk_aux_decl` as protected.

> I think the case here for having a separate API in Lean and in C++ is similar to the case for having private functions in some modules. They don't provide a good way to use the module and there's no good use cases that need them.

The C++ function `add_protected` is exported, and it is used from lots of other modules.  It very much is the public API.

I'm not going to argue whether this is a good or a bad API, but it's the one that we have and that is used everywhere.  In a sense it is not too different from attributes, which can also be applied after the fact:
```lean
def foo := nat
attribute [class] foo
attribute [irreducible] foo
```